bindings/java/fdbJNI.cpp

```diff
@@ -640,6 +640,150 @@ JNIEXPORT jlong JNICALL Java_com_apple_foundationdb_FDBTransaction_Transaction_1
+		offset += kvs[count - 1].key_length;
+	} else {
+		int zero = 0;
+		memcpy(buffer + offset, &zero, sizeof(jint));
+		offset += sizeof(jint);
+	}
+
+	for (int i = 0; i < count; i++) {
+		memcpy(buffer + offset, &kvs[i].key_length, sizeof(jint));
+		memcpy(buffer + offset + sizeof(jint), &kvs[i].value_length, sizeof(jint));
+		offset += 2 * sizeof(jint);
+
+		memcpy(buffer + offset, kvs[i].key, kvs[i].key_length);
+		offset += kvs[i].key_length;
+
+		memcpy(buffer + offset, kvs[i].value, kvs[i].value_length);
+		offset += kvs[i].value_length;
+	}
+}
+
+int readKeySelector(uint8_t* buffer, jint* numBytes, jint* orEqual, jint* offset, uint8_t** bytes) {
+	memcpy(numBytes, buffer, sizeof(jint));
+	memcpy(orEqual, buffer + sizeof(jint), sizeof(jint));
+	memcpy(offset, buffer + 2*sizeof(jint), sizeof(jint));
+	*bytes = buffer + sizeof(jint) * 3;
+	return *numBytes + sizeof(jint) * 3;
+}
+
+JNIEXPORT void JNICALL Java_com_apple_foundationdb_FDBTransaction_Transaction_1getDirectRange(
+	JNIEnv* jenv, jobject, jlong tPtr, jobject jbuffer, jint bufferCapacity) {
+
+	if (!tPtr || !jbuffer) {
+		throwParamNotNull(jenv);
+		return;
+	}
+
+	uint8_t* buffer = (uint8_t*)jenv->GetDirectBufferAddress(jbuffer);
+	if (!buffer) {
+		if (!jenv->ExceptionOccurred()) throwRuntimeEx(jenv, "Error getting handle to native resources");
+		return;
+	}
+
+	int readOffset = 0;
+
+	jint lengthBegin;
+	jint orEqualBegin;
+	jint offsetBegin;
+	uint8_t* barrBegin;
+	readOffset += readKeySelector(buffer + readOffset, &lengthBegin, &orEqualBegin, &offsetBegin, &barrBegin);
+
+	jint lengthEnd;
+	jint orEqualEnd;
+	jint offsetEnd;
+	uint8_t* barrEnd;
+	readOffset += readKeySelector(buffer + readOffset, &lengthEnd, &orEqualEnd, &offsetEnd, &barrEnd);
+
+	jint rowLimit;
+	memcpy(&rowLimit, buffer + readOffset, sizeof(jint));
+	readOffset += sizeof(jint);
+
+	jint targetBytes;
+	memcpy(&targetBytes, buffer + readOffset, sizeof(jint));
+	readOffset += sizeof(jint);
+
+	jint streamingMode;
+	memcpy(&streamingMode, buffer + readOffset, sizeof(jint));
+	readOffset += sizeof(jint);
+
+	jint iteration;
+	memcpy(&iteration, buffer + readOffset, sizeof(jint));
+	readOffset += sizeof(jint);
+
+	jint snapshot;
+	memcpy(&snapshot, buffer + readOffset, sizeof(jint));
+	readOffset += sizeof(jint);
+
+	jint reverse;
+	memcpy(&reverse, buffer + readOffset, sizeof(jint));
+	readOffset += sizeof(jint);
+
+
+	// Update targetBytes based on our buffer capacity
+	if (targetBytes < bufferCapacity - readOffset*2) {
+		// Rough slack to consider return metadata
+		targetBytes = bufferCapacity - readOffset*2;
+	}
+
+	FDBTransaction* tr = (FDBTransaction*)tPtr;
+
+	FDBFuture* f = fdb_transaction_get_range(tr, barrBegin, lengthBegin, orEqualBegin, offsetBegin, barrEnd, lengthEnd,
+											 orEqualEnd, offsetEnd, rowLimit, targetBytes,
+											 (FDBStreamingMode)streamingMode, iteration, snapshot, reverse);
+
+	if (fdb_future_block_until_ready(f) != 0) {
```

Comment:
Yeah, I've an implementation ready with that design as well. I will run some benchmarks for that. The reason I spent time doing this was because JNI was the hotspot in general in John's profiling (the original issue was around to reduce the JNI calls). So I started with reducing all the JNI calls and small memory allocations. Earlier, we were copying over begin/end bytes in newly allocated array and few more JNI calls.

But I agree we should profile further to look precisely where the actual cost is. Also, another thing I've to verify that when we cancel the future in Java world, it interrupts the JNI block. Because if it doesn't, it potentially means we'll still freeing up the DirectBuffer, which can be used by both old JNI call and potentially new JNI call.